packages/orchestration/src/util/stream.ts

```diff
@@ -0,0 +1,351 @@
+        (toolCall.function?.arguments ?? '');
+    } else {
+      // Add new tool call
+      mergedToolCalls.push(transformStreamingToolCall(toolCall));
+    }
+  }
+  return mergedToolCalls;
+}
+
+function mergeLogProbs(
+  existing: ChoiceLogprobs | undefined,
+  incoming: ChoiceLogprobs | undefined
+): ChoiceLogprobs | undefined {
+  if (!incoming) {
+    return existing;
+  }
+  if (!existing) {
+    return incoming;
+  }
+  return {
+    content: [...(existing.content ?? []), ...(incoming.content ?? [])],
+    refusal: [...(existing.refusal ?? []), ...(incoming.refusal ?? [])]
+  };
+}
+
+function handleFinishReason(
+  existing: string | undefined,
+  incoming: string | undefined,
+  choiceIndex: number | undefined
+): string {
+  if (!incoming) {
+    return existing ?? '';
+  }
+
+  switch (incoming) {
+    case 'content_filter':
+      logger.error(
+        `Choice ${choiceIndex}: Stream finished with content filter hit.`
+      );
+      break;
+    case 'length':
+      logger.error(
+        `Choice ${choiceIndex}: Stream finished with token length exceeded.`
+      );
+      break;
+    case 'stop':
+    case 'tool_calls':
+    case 'function_call':
+      logger.debug(`Choice ${choiceIndex}: Stream finished.`);
+      break;
+    default:
+      logger.error(
+        `Choice ${choiceIndex}: Stream finished with unknown reason '${incoming}'.`
+      );
+  }
+
+  return incoming;
+}
+
+function transformStreamingChoice(choice: LlmChoiceStreaming): LlmChoice {
+  return {
+    index: choice.index,
+    message: {
+      role: 'assistant',
+      content: choice.delta.content,
+      tool_calls: transformStreamingToolCalls(choice.delta.tool_calls),
+      refusal: choice.delta.refusal
+    },
+    finish_reason: choice.finish_reason ?? '',
+    logprobs: choice.logprobs
+  };
+}
+
+function transformStreamingToolCalls(
+  toolCalls: ToolCallChunk[] | undefined
+): MessageToolCall[] | undefined {
+  if (!toolCalls || toolCalls.length === 0) {
+    return undefined;
+  }
+  return toolCalls?.map(toolCall => transformStreamingToolCall(toolCall));
+}
+
+function transformStreamingToolCall(toolCall: ToolCallChunk): MessageToolCall {
+  return {
+    index: toolCall.index,
+    id: toolCall.id ?? '',
+    type: toolCall.type ?? 'function',
+    function: {
+      name: toolCall.function?.name ?? '',
+      arguments: toolCall.function?.arguments ?? ''
+    }
+  };
+}
+
+/**
+ * @internal
+ */
+export function validateResponse(
+  response: OrchestrationStreamResponse<OrchestrationStreamChunkResponse>
+): void {
+  if (response._openStream) {
+    throw new Error(
+      "Stream wasn't closed properly. Please ensure the stream is closed after processing."
+    );
+  }
+
+  validateLlmModuleResult(response._data.module_results?.llm);
+
+  validateLlmModuleResult(response._data.orchestration_result);
+
+  validateChoices(response._data.module_results?.output_unmasking);
+}
+
+function validateLlmModuleResult(
```

Comment:
[pp] Not sure if this is always required for everything, but consider adding the source `module_results` or `orchestration_result` to logs to know the exact property being logged out.

Reply:
Good idea